### DISCUSSION TITLE
Custom make rules

### DISCUSSION THREAD

**Discussion by holychicken99** [premake/premake-core#1707] [Q&A] [answered]
Hi,
I am new to premake and had a doubt about custom rules for makefile generated by 
``` bash
premake5 gmake2
```
is there a way to add a custom rule which can later be invoked by saying 

``` bash 
make {something}
```

**Comment by starkos** [comment, marked as answer]
> is there a way to add a custom rule which can later be invoked by saying

No, that's not supported. The custom build rules are more like "when see files of type X, here are the build steps to apply to them", and are intended for things like shaders and DSLs. It sounds like you're asking for something more like a custom makefile target, which you can actually inject into your generated makefiles using [`makesettings()`](https://premake.github.io/docs/makesettings/).

Does that help?

**Reply by holychicken99** [reply]
hey , thx for the response !
I read the docs for makesettings and  tried to do this 

``` bash
makesettings[[
    foo = ok

    all:
    ifeq ($(foo), ok)
        echo "foo equals ok"
    else
        echo "nope"
    endif
]]
```
this did not work as I was getting 
``` bash 
==== Building output (debug) ====
output.make:34: *** missing separator.  Stop.
make: *** [Makefile:30: output] Error 2

```
also why does the shown example use [[ ]] ?

Thanks in advance !

**Comment by starkos** [comment]
Have you opened up the generated makefile and tried to investigate the issue at line 34? You probably shouldn't be defining an `all:` target since Premake has already create one for you, and makefile file targets should not be indented.

> why does the shown example use [[ ]] ?

See [String Tutorial on lua-users.org](http://lua-users.org/wiki/StringsTutorial).

**Comment by holychicken99** [comment]
I have a test directory which contains a test.cpp file and I wanted to define a premake configuration such that the generated  makefile 
will let me write 
``` bash
make test 
```
and will then compile the test.cpp file. Thats why I was thinking of adding a custom target test. what should I do in this case ?
thank you

**Reply by samsinsane** [reply]
I'm pretty sure you can achieve `make test` using:
```lua
project "test"
  files { "tests/**.cpp" }
```
I've done it once before with a Premake generated makefile, so I could be wrong on exactly how it works but I'm pretty sure that's it.
